<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title>Publications | Dhananjay Bhaskar</title> <meta name="author" content="Dhananjay Bhaskar"> <meta name="description" content="Personal academic webpage for Dhananjay Bhaskar based on [*folio](https://github.com/bogoli/-folio) design. "> <meta name="keywords" content="dhananjay, bhaskar, academic-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.css"> <link rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons"> <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?19f3075a2d19613090fe9e16b564e1fe" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="/assets/img/favicon.ico?bca6fdf3cfe0a8634226215d7c159293"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://dbhaskar92.github.io/publications/"> <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?e74e74bf055e5729d44a7d031a5ca6a5" media="none" id="highlight_theme_dark"> <script src="/assets/js/theme.js?96d6b3e1c3604aca8b6134c7afdd5db6"></script> <script src="/assets/js/dark_mode.js?9b17307bb950ffa2e34be0227f53558f"></script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"><span class="font-weight-bold">Dhananjay </span>Bhaskar</a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">About</a> </li> <li class="nav-item active"> <a class="nav-link" href="/publications/">Publications<span class="sr-only">(current)</span></a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">CV</a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="fas fa-moon"></i> <i class="fas fa-sun"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5"> <div class="post"> <header class="post-header"> <h1 class="post-title">Publications</h1> <p class="post-description"></p> </header> <article> <h2 id="journal-articles"><strong>Journal Articles</strong></h2> <div class="publications"> <h2 class="bibliography">2024</h2> <ol class="bibliography"> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">IEEE TOSP</abbr></div> <div id="tong_learnable_2024" class="col-sm-8"> <div class="title">Learnable Filters for Geometric Scattering Modules</div> <div class="author"> Alexander Tong, Frederik Wenkel, <em>Dhananjay Bhaskar</em>, Kincaid Macdonald, Jackson Grady, Michael Perlmutter, Smita Krishnaswamy, and Guy Wolf</div> <div class="periodical"> <em>IEEE Transactions on Signal Processing</em>, 2024 </div> <div class="periodical"> Conference Name: IEEE Transactions on Signal Processing </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> <a href="https://github.com/KrishnaswamyLab/LearnableScattering" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">Code</a> </div> <div class="abstract hidden"> <p>We propose a new graph neural network (GNN) module, based on relaxations of recently proposed geometric scattering transforms, which consist of a cascade of graph wavelet filters. Our learnable geometric scattering (LEGS) module enables adaptive tuning of the wavelets to encourage band-pass features to emerge in learned representations. The incorporation of our LEGS-module in GNNs enables the learning of longer-range graph relations compared to many popular GNNs, which often rely on encoding graph structure via smoothness or similarity between neighbors. Further, its wavelet priors result in simplified architectures with significantly fewer learned parameters compared to competing GNNs. We demonstrate the predictive performance of LEGS-based networks on graph classification benchmarks, as well as the descriptive quality of their learned features in biochemical graph data exploration tasks. Our results show that LEGS-based networks match or outperforms popular GNNs, as well as the original geometric scattering construction, on many datasets, in particular in biochemical domains, while retaining certain mathematical properties of handcrafted (non-learned) geometric scattering.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">Sci Rep</abbr></div> <div id="manrique-castano_dissecting_2024" class="col-sm-8"> <div class="title">Dissecting glial scar formation by spatial point pattern and topological data analysis</div> <div class="author"> Daniel Manrique-Castano, <em>Dhananjay Bhaskar</em>, and Ayman ElAli</div> <div class="periodical"> <em>Scientific Reports</em>, Aug 2024 </div> <div class="periodical"> Publisher: Nature Publishing Group </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> <a href="https://github.com/elalilab/Stroke_GlialScar_PPA-TDA" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">Code</a> </div> <div class="abstract hidden"> <p>Glial scar formation represents a fundamental response to central nervous system (CNS) injuries. It is mainly characterized by a well-defined spatial rearrangement of reactive astrocytes and microglia. The mechanisms underlying glial scar formation have been extensively studied, yet quantitative descriptors of the spatial arrangement of reactive glial cells remain limited. Here, we present a novel approach using point pattern analysis (PPA) and topological data analysis (TDA) to quantify spatial patterns of reactive glial cells after experimental ischemic stroke in mice. We provide open and reproducible tools using R and Julia to quantify spatial intensity, cell covariance and conditional distribution, cell-to-cell interactions, and short/long-scale arrangement, which collectively disentangle the arrangement patterns of the glial scar. This approach unravels a substantial divergence in the distribution of GFAP+ and IBA1+ cells after injury that conventional analysis methods cannot fully characterize. PPA and TDA are valuable tools for studying the complex spatial arrangement of reactive glia and other nervous cells following CNS injuries and have potential applications for evaluating glial-targeted restorative therapies.</p> </div> </div> </div> </li> </ol> <h2 class="bibliography">2023</h2> <ol class="bibliography"> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">JCB</abbr></div> <div id="moore_cell_2023" class="col-sm-8"> <div class="title">Cell cycle controls long-range calcium signaling in the regenerating epidermis</div> <div class="author"> Jessica L. Moore, <em>Dhananjay Bhaskar</em>, Feng Gao, Catherine Matte-Martone, Shuangshuang Du, Elizabeth Lathrop, Smirthy Ganesan, Lin Shao, Rachael Norris, Nil Campamà Sanz, Karl Annusver, Maria Kasper, Andy Cox, Caroline Hendry, Bastian Rieck, Smita Krishnaswamy, and Valentina Greco</div> <div class="periodical"> <em>The Journal of Cell Biology</em>, Jul 2023 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Skin homeostasis is maintained by stem cells, which must communicate to balance their regenerative behaviors. Yet, how adult stem cells signal across regenerative tissue remains unknown due to challenges in studying signaling dynamics in live mice. We combined live imaging in the mouse basal stem cell layer with machine learning tools to analyze patterns of Ca2+ signaling. We show that basal cells display dynamic intercellular Ca2+ signaling among local neighborhoods. We find that these Ca2+ signals are coordinated across thousands of cells and that this coordination is an emergent property of the stem cell layer. We demonstrate that G2 cells are required to initiate normal levels of Ca2+ signaling, while connexin43 connects basal cells to orchestrate tissue-wide coordination of Ca2+ signaling. Lastly, we find that Ca2+ signaling drives cell cycle progression, revealing a communication feedback loop. This work provides resolution into how stem cells at different cell cycle stages coordinate tissue-wide signaling during epidermal regeneration.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">NPJ</abbr></div> <div id="bhaskar_topological_2023" class="col-sm-8"> <div class="title">Topological data analysis of spatial patterning in heterogeneous cell populations: clustering and sorting with varying cell-cell adhesion</div> <div class="author"> <em>Dhananjay Bhaskar</em>, William Y. Zhang, Alexandria Volkening, Björn Sandstede, and Ian Y. Wong</div> <div class="periodical"> <em>npj Systems Biology and Applications</em>, Sep 2023 </div> <div class="periodical"> Number: 1 Publisher: Nature Publishing Group </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Different cell types aggregate and sort into hierarchical architectures during the formation of animal tissues. The resulting spatial organization depends (in part) on the strength of adhesion of one cell type to itself relative to other cell types. However, automated and unsupervised classification of these multicellular spatial patterns remains challenging, particularly given their structural diversity and biological variability. Recent developments based on topological data analysis are intriguing to reveal similarities in tissue architecture, but these methods remain computationally expensive. In this article, we show that multicellular patterns organized from two interacting cell types can be efficiently represented through persistence images. Our optimized combination of dimensionality reduction via autoencoders, combined with hierarchical clustering, achieved high classification accuracy for simulations with constant cell numbers. We further demonstrate that persistence images can be normalized to improve classification for simulations with varying cell numbers due to proliferation. Finally, we systematically consider the importance of incorporating different topological features as well as information about each cell type to improve classification accuracy. We envision that topological machine learning based on persistence images will enable versatile and robust classification of complex tissue architectures that occur in development and disease.</p> </div> </div> </div> </li> </ol> <h2 class="bibliography">2022</h2> <ol class="bibliography"><li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">Nature</abbr></div> <div id="castro_transformer-based_2022" class="col-sm-8"> <div class="title">Transformer-based protein generation with regularized latent space optimization</div> <div class="author"> Egbert Castro, Abhinav Godavarthi, Julian Rubinfien, Kevin Givechian, <em>Dhananjay Bhaskar</em>, and Smita Krishnaswamy</div> <div class="periodical"> <em>Nature Machine Intelligence</em>, Oct 2022 </div> <div class="periodical"> Number: 10 Publisher: Nature Publishing Group </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>The development of powerful natural language models has improved the ability to learn meaningful representations of protein sequences. In addition, advances in high-throughput mutagenesis, directed evolution and next-generation sequencing have allowed for the accumulation of large amounts of labelled fitness data. Leveraging these two trends, we introduce Regularized Latent Space Optimization (ReLSO), a deep transformer-based autoencoder, which features a highly structured latent space that is trained to jointly generate sequences as well as predict fitness. Through regularized prediction heads, ReLSO introduces a powerful protein sequence encoder and a novel approach for efficient fitness landscape traversal. Using ReLSO, we explicitly model the sequence–function landscape of large labelled datasets and generate new molecules by optimizing within the latent space using gradient-based methods. We evaluate this approach on several publicly available protein datasets, including variant sets of anti-ranibizumab and green fluorescent protein. We observe a greater sequence optimization efficiency (increase in fitness per optimization step) using ReLSO compared with other approaches, where ReLSO more robustly generates high-fitness sequences. Furthermore, the attention-based relationships learned by the jointly trained ReLSO models provide a potential avenue towards sequence-level fitness attribution information.</p> </div> </div> </div> </li></ol> <h2 class="bibliography">2021</h2> <ol class="bibliography"><li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">RSC</abbr></div> <div id="bhaskar_topological_2021" class="col-sm-8"> <div class="title">Topological data analysis of collective and individual epithelial cells using persistent homology of loops</div> <div class="author"> <em>Dhananjay Bhaskar</em>, William Y. Zhang, and Ian Y. Wong</div> <div class="periodical"> <em>Soft Matter</em>, May 2021 </div> <div class="periodical"> Publisher: The Royal Society of Chemistry </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8276269/pdf/nihms-1718172.pdf" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">PDF</a> </div> <div class="abstract hidden"> <p>Interacting, self-propelled particles such as epithelial cells can dynamically self-organize into complex multicellular patterns, which are challenging to classify without a priori information. Classically, different phases and phase transitions have been described based on local ordering, which may not capture structural features at larger length scales. Instead, topological data analysis (TDA) determines the stability of spatial connectivity at varying length scales (i.e. persistent homology), and can compare different particle configurations based on the “cost” of reorganizing one configuration into another. Here, we demonstrate a topology-based machine learning approach for unsupervised profiling of individual and collective phases based on large-scale loops. We show that these topological loops (i.e. dimension 1 homology) are robust to variations in particle number and density, particularly in comparison to connected components (i.e. dimension 0 homology). We use TDA to map out phase diagrams for simulated particles with varying adhesion and propulsion, at constant population size as well as when proliferation is permitted. Next, we use this approach to profile our recent experiments on the clustering of epithelial cells in varying growth factor conditions, which are compared to our simulations. Finally, we characterize the robustness of this approach at varying length scales, with sparse sampling, and over time. Overall, we envision TDA will be broadly applicable as a model-agnostic approach to analyze active systems with varying population size, from cytoskeletal motors to motile cells to flocking or swarming animals.</p> </div> </div> </div> </li></ol> <h2 class="bibliography">2019</h2> <ol class="bibliography"> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">AIP</abbr></div> <div id="bhaskar_analyzing_2019" class="col-sm-8"> <div class="title">Analyzing collective motion with machine learning and topology</div> <div class="author"> <em>Dhananjay Bhaskar</em>, Angelika Manhart, Jesse Milzman, John T. Nardini, Kathleen M. Storey, Chad M. Topaz, and Lori Ziegelmeier</div> <div class="periodical"> <em>Chaos: An Interdisciplinary Journal of Nonlinear Science</em>, Dec 2019 </div> <div class="periodical"> Publisher: American Institute of Physics </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> <a href="https://github.com/dbhaskar92/AMS-MRC-2018" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">Code</a> </div> <div class="abstract hidden"> <p>We use topological data analysis and machine learning to study a seminal model of collective motion in biology [M. R. D’Orsogna et al., Phys. Rev. Lett. 96, 104302 (2006)]. This model describes agents interacting nonlinearly via attractive-repulsive social forces and gives rise to collective behaviors such as flocking and milling. To classify the emergent collective motion in a large library of numerical simulations and to recover model parameters from the simulation data, we apply machine learning techniques to two different types of input. First, we input time series of order parameters traditionally used in studies of collective motion. Second, we input measures based on topology that summarize the time-varying persistent homology of simulation data over multiple scales. This topological approach does not require prior knowledge of the expected patterns. For both unsupervised and supervised machine learning methods, the topological approach outperforms the one that is based on traditional order parameters.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">PNAS</abbr></div> <div id="leggett_motility-limited_2019" class="col-sm-8"> <div class="title">Motility-limited aggregation of mammary epithelial cells into fractal-like clusters</div> <div class="author"> Susan E. Leggett, Zachary J. Neronha, <em>Dhananjay Bhaskar</em>, Jea Yun Sim, Theodora Myrto Perdikari, and Ian Y. Wong</div> <div class="periodical"> <em>Proceedings of the National Academy of Sciences of the United States of America</em>, Aug 2019 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Migratory cells transition between dispersed individuals and multicellular collectives during development, wound healing, and cancer. These transitions are associated with coordinated behaviors as well as arrested motility at high cell densities, but remain poorly understood at lower cell densities. Here, we show that dispersed mammary epithelial cells organize into arrested, fractal-like clusters at low density in reduced epidermal growth factor (EGF). These clusters exhibit a branched architecture with a fractal dimension of [Formula: see text], reminiscent of diffusion-limited aggregation of nonliving colloidal particles. First, cells display diminished motility in reduced EGF, which permits irreversible adhesion upon cell-cell contact. Subsequently, leader cells emerge that guide collectively migrating strands and connect clusters into space-filling networks. Thus, this living system exhibits gelation-like arrest at low cell densities, analogous to the glass-like arrest of epithelial monolayers at high cell densities. We quantitatively capture these behaviors with a jamming-like phase diagram based on local cell density and EGF. These individual to collective transitions represent an intriguing link between living and nonliving systems, with potential relevance for epithelial morphogenesis into branched architectures.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">ACS</abbr></div> <div id="khoo_breast_2019" class="col-sm-8"> <div class="title">Breast Cancer Cells Transition from Mesenchymal to Amoeboid Migration in Tunable Three-Dimensional Silk-Collagen Hydrogels</div> <div class="author"> Amanda S. Khoo, Thomas M. Valentin, Susan E. Leggett, <em>Dhananjay Bhaskar</em>, Elisa M. Bye, Shoham Benmelech, Blanche C. Ip, and Ian Y. Wong</div> <div class="periodical"> <em>ACS biomaterials science &amp; engineering</em>, Sep 2019 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Invading cancer cells adapt their migration phenotype in response to mechanical and biochemical cues from the extracellular matrix. For instance, mesenchymal migration is associated with strong cell-matrix adhesions and an elongated morphology, while amoeboid migration is associated with minimal cell-matrix adhesions and a rounded morphology. However, it remains challenging to elucidate the role of matrix mechan-ics and biochemistry, since these are both dependent on ECM protein concentration. Here, we demonstrate a composite silk fibroin and collagen I hydrogel where stiffness and microstructure can be systematically tuned over a wide range. Using an overlay assay geometry, we show that the invasion of metastatic breast cancer cells exhibits a biphasic dependence on silk fibroin concentration at fixed collagen I concentration, first increasing as the hydrogel stiffness increases, then decreasing as the pore size of silk fibroin decreases. Indeed, mesenchymal morphology exhibits a similar biphasic depen-dence on silk fibroin concentration, while amoeboid morphologies were favored when cell-matrix adhesions were less effective. We used exogenous biochemical treatment to perturb cells towards increased contractility and a mesenchymal morphology, as well as to disrupt cytoskeletal function and promote an amoeboid morphology. Overall, we envision that this tunable biomaterial platform in a 96-well plate format will be widely applicable to screen cancer cell migration against combinations of designer biomaterials and targeted inhibitors.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">RSC</abbr></div> <div id="valentin_3d_2019" class="col-sm-8"> <div class="title">3D printed self-adhesive PEGDA–PAA hydrogels as modular components for soft actuators and microfluidics</div> <div class="author"> Thomas M. Valentin, Eric M. DuBois, Catherine E. Machnicki, <em>Dhananjay Bhaskar</em>, Francis R. Cui, and Ian Y. Wong</div> <div class="periodical"> <em>Polymer Chemistry</em>, Apr 2019 </div> <div class="periodical"> Publisher: The Royal Society of Chemistry </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Hydrogel building blocks that are stimuli-responsive and self-adhesive could be utilized as a simple “do-it-yourself” construction set for soft machines and microfluidic devices. However, conventional covalently-crosslinked hydrogels are unsuitable since they are as static materials with poor interfacial adhesion. In this article, we demonstrate ion-responsive interchangeable parts based on composite hydrogels that incorporate both covalent and ionic crosslinking. We use light-directed 3D printing to covalently-crosslink poly(ethylene glycol) diacrylate in the presence of anionic poly(acrylic acid) of much higher molecular weight. The addition of trivalent cations acts to crosslink the anionic polymer chains together. Using high cation concentrations drives strong crosslinking, which can result in dramatic hydrogel contraction. Mismatched contraction of layered ion-responsive and non-ion-responsive hydrogels can control bending and twisting actuation, which is utilized for a gripping device. Alternatively, moderate cation concentrations permit strong self-adhesion between hydrogel surfaces. LEGO-like hydrogel blocks with internal channels and external mechanical connectors can be stacked into complex microfluidic device geometries including serpentine micromixers and multilevel architectures. This approach enables “plug-and-play” hydrogel parts for ionic soft machines that mimic actuation, sensing, and fluid transport in living systems.</p> </div> </div> </div> </li> </ol> <h2 class="bibliography">2018</h2> <ol class="bibliography"><li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">Phys. Biol.</abbr></div> <div id="zmurchok_coupling_2018" class="col-sm-8"> <div class="title">Coupling mechanical tension and GTPase signaling to generate cell and tissue dynamics</div> <div class="author"> Cole Zmurchok, <em>Dhananjay Bhaskar</em>, and Leah Edelstein-Keshet</div> <div class="periodical"> <em>Physical Biology</em>, Apr 2018 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Regulators of the actin cytoskeleton such Rho GTPases can modulate forces developed in cells by promoting actomyosin contraction. At the same time, through mechanosensing, tension is known to affect the activity of Rho GTPases. What happens when these effects act in concert? Using a minimal model (1 GTPase coupled to a Kelvin–Voigt element), we show that two-way feedback between signaling (‘RhoA’) and mechanical tension (stretching) leads to a spectrum of cell behaviors, including contracted or relaxed cells, and cells that oscillate between these extremes. When such ‘model cells’ are connected to one another in a row or in a 2D sheet (‘epithelium’), we observe waves of contraction/relaxation and GTPase activity sweeping through the tissue. The minimal model lends itself to full bifurcation analysis, and suggests a mechanism that explains behavior observed in the context of development and collective cell behavior.</p> </div> </div> </div> </li></ol> <h2 class="bibliography">2017</h2> <ol class="bibliography"><li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">PLOS</abbr></div> <div id="knutsdottir_polarization_2017" class="col-sm-8"> <div class="title">Polarization and migration in the zebrafish posterior lateral line system</div> <div class="author"> Hildur Knutsdottir, Cole Zmurchok, <em>Dhananjay Bhaskar</em>, Eirikur Palsson, Damian Dalle Nogare, Ajay B. Chitnis, and Leah Edelstein-Keshet</div> <div class="periodical"> <em>PLOS Computational Biology</em>, Apr 2017 </div> <div class="periodical"> Publisher: Public Library of Science </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Collective cell migration plays an important role in development. Here, we study the posterior lateral line primordium (PLLP) a group of about 100 cells, destined to form sensory structures, that migrates from head to tail in the zebrafish embryo. We model mutually inhibitory FGF-Wnt signalling network in the PLLP and link tissue subdivision (Wnt receptor and FGF receptor activity domains) to receptor-ligand parameters. We then use a 3D cell-based simulation with realistic cell-cell adhesion, interaction forces, and chemotaxis. Our model is able to reproduce experimentally observed motility with leading cells migrating up a gradient of CXCL12a, and trailing (FGF receptor active) cells moving actively by chemotaxis towards FGF ligand secreted by the leading cells. The 3D simulation framework, combined with experiments, allows an investigation of the role of cell division, chemotaxis, adhesion, and other parameters on the shape and speed of the PLLP. The 3D model demonstrates reasonable behaviour of control as well as mutant phenotypes.</p> </div> </div> </div> </li></ol> </div> <hr> <h2 id="peer-reviewed-cs-conferences"><strong>Peer-Reviewed CS Conferences</strong></h2> <div class="publications"> <h2 class="bibliography">2024</h2> <ol class="bibliography"><li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">LoG</abbr></div> <div id="bhaskar_inferring_2024" class="col-sm-8"> <div class="title">Inferring Dynamic Regulatory Interaction Graphs From Time Series Data With Perturbations</div> <div class="author"> <em>Dhananjay Bhaskar</em>, Daniel Sumner Magruder, Matheo Morales, Edward De Brouwer, Aarthi Venkat, Frederik Wenkel, James Noonan, Guy Wolf, Natalia Ivanova, and Smita Krishnaswamy</div> <div class="periodical"> <em>In Proceedings of the Second Learning on Graphs Conference</em>, Apr 2024 </div> <div class="periodical"> ISSN: 2640-3498 </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Complex systems are characterized by intricate interactions between entities that evolve dynamically over time. Accurate inference of these dynamic relationships is crucial for understanding and predicting system behavior. In this paper, we propose Regulatory Temporal Interaction Network Inference (RiTINI) for inferring time-varying interaction graphs in complex systems using a novel combination of space-and-time graph attentions and graph neural ordinary differential equations (ODEs). RiTINI leverages time-lapse signals on a graph prior, as well as perturbations of signals at various nodes in order to effectively capture the dynamics of the underlying system. This approach is distinct from traditional causal inference networks, which are limited to inferring acyclic and static graphs. In contrast, RiTINI can infer cyclic, directed, and time-varying graphs, providing a more comprehensive and accurate representation of complex systems. The graph attention mechanism in RiTINI allows the model to adaptively focus on the most relevant interactions in time and space, while the graph neural ODEs enable continuous-time modeling of the system’s dynamics. We evaluate RiTINI’s performance on simulations of dynamical systems, neuronal networks, and gene regulatory networks, demonstrating its state-of-the-art capability in inferring interaction graphs compared to previous methods.</p> </div> </div> </div> </li></ol> <h2 class="bibliography">2023</h2> <ol class="bibliography"><li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">IEEE MLSP</abbr></div> <div id="macdonald_flow_2023" class="col-sm-8"> <div class="title">A Flow Artist for High-Dimensional Cellular Data</div> <div class="author"> Kincaid MacDonald, <em>Dhananjay Bhaskar</em>, Guy Thampakkul, Nhi Nguyen, Joia Zhang, Michael Perlmutter, Ian Adelstein, and Smita Krishnaswamy</div> <div class="periodical"> <em>In 2023 IEEE 33rd International Workshop on Machine Learning for Signal Processing (MLSP)</em>, Apr 2023 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>We consider the problem of embedding point cloud data sampled from an underlying manifold with an associated flow or velocity. Such data arises in many contexts where static snapshots of dynamic entities are measured, including in high-throughput biology such as single-cell transcriptomics. Existing embedding techniques either do not utilize velocity information or embed the coordinates and velocities independently, i.e., they either impose velocities on top of an existing point embedding or embed points within a prescribed vector field. Here we present FlowArtist, a neural network that embeds points while jointly learning a vector field around the points. The combination allows FlowArtist to better separate and visualize velocity-informed structures. Our results, on toy datasets and single-cell RNA velocity data, illustrate the value of utilizing coordinate and velocity information in tandem for embedding and visualizing high-dimensional data.</p> </div> </div> </div> </li></ol> <h2 class="bibliography">2022</h2> <ol class="bibliography"> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">IEEE MLSP</abbr></div> <div id="bhaskar_molecular_2022" class="col-sm-8"> <div class="title">Molecular Graph Generation via Geometric Scattering</div> <div class="author"> <em>Dhananjay Bhaskar</em>, Jackson Grady, Egbert Castro, Michael Perlmutter, and Smita Krishnaswamy</div> <div class="periodical"> <em>In 2022 IEEE 32nd International Workshop on Machine Learning for Signal Processing (MLSP)</em>, Aug 2022 </div> <div class="periodical"> ISSN: 2161-0371 </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Although existing deep learning models perform remarkably well at predicting physicochemical properties and binding affinities, the generation of new molecules with optimized properties remains challenging. Inherently, most GNNs perform poorly in whole-graph representation due to the limitations of the message-passing paradigm. Furthermore, step-by-step graph generation frameworks that use reinforcement learning or other sequential processing can be slow and result in a high proportion of invalid molecules with substantial post-processing needed in order to generate valid molecules. To address these issues, we propose a representation-first approach to molecular graph generation. We guide the latent representation of an autoencoder by capturing graph structure information with the geometric scattering transform and apply penalties that organize the representation by molecular properties. We show that this highly structured latent space can be directly used for molecular graph generation by the use of a GAN. We demonstrate that our architecture learns meaningful representations of drug datasets and provides a platform for drug synthesis using publicly available ZINC and BindingDB datasets.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">NeurIPS</abbr></div> <div id="bhaskar_diffusion_2022" class="col-sm-8"> <div class="title">Diffusion Curvature for Estimating Local Curvature in High Dimensional Data</div> <div class="author"> <em>Dhananjay Bhaskar</em>, Kincaid MacDonald, Oluwadamilola Fasina, Dawson Thomas, Bastian Rieck, Ian Adelstein, and Smita Krishnaswamy</div> <div class="periodical"> <em>Advances in Neural Information Processing Systems</em>, Dec 2022 </div> <div class="periodical"> </div> <div class="links"> </div> </div> </div> </li> </ol> </div> <hr> <h2 id="reviews"><strong>Reviews</strong></h2> <div class="publications"> <h2 class="bibliography">2023</h2> <ol class="bibliography"><li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">Review</abbr></div> <div id="venkat_multiscale_2023" class="col-sm-8"> <div class="title">Multiscale geometric and topological analyses for characterizing and predicting immune responses from single cell data</div> <div class="author"> Aarthi Venkat, <em>Dhananjay Bhaskar</em>, and Smita Krishnaswamy</div> <div class="periodical"> <em>Trends in Immunology</em>, Jul 2023 </div> <div class="periodical"> Publisher: Elsevier </div> <div class="links"> </div> </div> </div> </li></ol> <h2 class="bibliography">2022</h2> <ol class="bibliography"> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">Review</abbr></div> <div id="bhaskar_current_2022" class="col-sm-8"> <div class="title">Current trends in artificial intelligence in reproductive endocrinology</div> <div class="author"> <em>Dhananjay Bhaskar</em>, T. Arthur Chang, and Shunping Wang</div> <div class="periodical"> <em>Current Opinion in Obstetrics &amp; Gynecology</em>, Aug 2022 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>In this review, we discuss various applications of artificial intelligence in different areas of reproductive medicine. We summarize the current findings with their potentials and limitations, and also discuss the future direction for research and clinical applications.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">Review</abbr></div> <div id="bhaskar_need_2022" class="col-sm-8"> <div class="title">The need for speed: Migratory cells in tight spaces boost their molecular clock</div> <div class="author"> <em>Dhananjay Bhaskar</em>, Alex M. Hruska, and Ian Y. Wong</div> <div class="periodical"> <em>Cell Systems</em>, Jul 2022 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Cells migrating in spatial confinement exhibit higher intracellular calcium levels, which increases the oscillation frequency of a “molecular clock” that synchronizes guanine nucleotide exchange factor GEF-H1 and microtubule polymerization for more frequent bursts of speed.</p> </div> </div> </div> </li> </ol> </div> <hr> <h2 id="preprints"><strong>Preprints</strong></h2> <div class="publications"> <h2 class="bibliography">2024</h2> <ol class="bibliography"> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">arXiv</abbr></div> <div id="viswanath_protscape_2024" class="col-sm-8"> <div class="title">ProtSCAPE: Mapping the landscape of protein conformations in molecular dynamics</div> <div class="author"> Siddharth Viswanath, <em>Dhananjay Bhaskar</em>, David R. Johnson, Joao Felipe Rocha, Egbert Castro, Jackson D. Grady, Alex T. Grigas, Michael A. Perlmutter, Corey S. O’Hern, and Smita Krishnaswamy</div> <div class="periodical"> Oct 2024 </div> <div class="periodical"> arXiv:2410.20317 </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Understanding the dynamic nature of protein structures is essential for comprehending their biological functions. While significant progress has been made in predicting static folded structures, modeling protein motions on microsecond to millisecond scales remains challenging. To address these challenges, we introduce a novel deep learning architecture, Protein Transformer with Scattering, Attention, and Positional Embedding (ProtSCAPE), which leverages the geometric scattering transform alongside transformer-based attention mechanisms to capture protein dynamics from molecular dynamics (MD) simulations. ProtSCAPE utilizes the multi-scale nature of the geometric scattering transform to extract features from protein structures conceptualized as graphs and integrates these features with dual attention structures that focus on residues and amino acid signals, generating latent representations of protein trajectories. Furthermore, ProtSCAPE incorporates a regression head to enforce temporally coherent latent representations.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">bioRxiv</abbr></div> <div id="bhaskar_neuro-gsth_2024" class="col-sm-8"> <div class="title">Neuro-GSTH: A Geometric Scattering and Persistent Homology Framework for Uncovering Spatiotemporal Signatures in Neural Activity</div> <div class="author"> <em>Dhananjay Bhaskar</em>, Jessica Moore, Yanlei Zhang, Feng Gao, Bastian Rieck, Guy Wolf, Helen Pushkarskaya, Firas Khasawneh, Elizabeth Munch, Valentina Greco, Christopher Pittenger, and Smita Krishnaswamy</div> <div class="periodical"> Oct 2024 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Understanding how neurons communicate and coordinate their activity is essential for unraveling the brain’s complex functionality. To analyze the intricate spatiotemporal dynamics of neural signaling, we developed Geometric Scattering Trajectory Homology (neuro-GSTH), a novel framework that captures time-evolving neural signals and encodes them into low-dimensional representations. GSTH integrates geometric scattering transforms, which extract multiscale features from brain signals modeled on anatomical graphs, with t-PHATE, a manifold learning method that maps the temporal evolution of neural activity. Topological descriptors from computational homology are then applied to characterize the global structure of these neural trajectories, enabling the quantification and differentiation of spatiotemporal brain dynamics. We demonstrate the power of neuro-GSTH in neuroscience by applying it to both simulated and biological neural datasets. First, we used neuro-GSTH to analyze neural oscillatory behavior in the Kuramoto model, revealing its capacity to track the synchronization of neural circuits as coupling strength increases. Next, we applied neuro-GSTH to neural recordings from the visual cortex of mice, where it accurately reconstructed visual stimulus patterns such as sinusoidal gratings. Neuro-GSTH-derived neural trajectories enabled precise classification of stimulus properties like spatial frequency and orientation, significantly outperforming traditional methods in capturing the underlying neural dynamics. These findings demonstrate that neuro-GSTH effectively identifies neural motifs—distinct patterns of spatiotemporal activity—providing a powerful tool for decoding brain activity across diverse tasks, sensory inputs, and neurological disorders. Neuro-GSTH thus offers new insights into neural communication and dynamics, advancing our ability to map and understand complex brain functions.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">bioRxiv</abbr></div> <div id="winn-nunez_generative_2024" class="col-sm-8"> <div class="title">Generative modeling of biological shapes and images using a probabilistic α-shape sampler</div> <div class="author"> Emily T. Winn-Nuñez, Hadley Witt, <em>Dhananjay Bhaskar</em>, Ryan Y. Huang, Jonathan S. Reichner, Ian Y. Wong, and Lorin Crawford</div> <div class="periodical"> Jan 2024 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Understanding morphological variation is an important task in many areas of computational biology. Recent studies have focused on developing computational tools for the task of sub-image selection which aims at identifying structural features that best describe the variation between classes of shapes. A major part in assessing the utility of these approaches is to demonstrate their performance on both simulated and real datasets. However, when creating a model for shape statistics, real data can be difficult to access and the sample sizes for these data are often small due to them being expensive to collect. Meanwhile, the current landscape of generative models for shapes has been mostly limited to approaches that use black-box inference—making it difficult to systematically assess the power and calibration of sub-image models. In this paper, we introduce the α-shape sampler: a probabilistic framework for generating realistic 2D and 3D shapes based on probability distributions which can be learned from real data. We demonstrate our framework using proof-of-concept examples and in two real applications in biology where we generate (i) 2D images of healthy and septic neutrophils and (ii) 3D computed tomography (CT) scans of primate mandibular molars. The α-shape sampler R package is open-source and can be downloaded at https://github.com/lcrawlab/ashapesampler., Using shapes and images to understand genotypic and phenotypic variation has proven to be an effective strategy in many biological applications. Unfortunately, shape data can be expensive to collect and, as a result, sample sizes for analyses are often small. Despite methodological advancements in shape statistics and machine learning, benchmarking standards for evaluating new computational tools via data simulation is still underdeveloped. In this paper, we present a probability-based pipeline called the α-shape sampler which has the flexibility to generate new and unobserved shapes based on an input set of data. We extensively evaluate the generative capabilities of our pipeline using 2D cellular images of neutrophils and 3D mandibular molars from two different suborders of primates.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">bioRxiv</abbr></div> <div id="koonce_neuroscan_2024" class="col-sm-8"> <div class="title">NeuroSCAN: Exploring Neurodevelopment via Spatiotemporal Collation of Anatomical Networks</div> <div class="author"> Noelle L. Koonce, Sarah E. Emerson, <em>Dhananjay Bhaskar</em>, Manik Kuchroo, Mark W. Moyle, Pura Arroyo-Morales, Nabor Vázquez Martínez, Smita Krishnaswamy, William Mohler, and Daniel Colón-Ramos</div> <div class="periodical"> Aug 2024 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Volume electron microscopy (vEM) datasets such as those generated for connectome studies allow nanoscale quantifications and comparisons of the cell biological features underpinning circuit architectures. Quantifications of cell biological relationships in the connectome result in rich multidimensional datasets that benefit from data science approaches, including dimensionality reduction and integrated graphical representations of neuronal relationships. We developed NeuroSCAN, an online open- source platform that bridges sophisticated graph analytics from data science approaches with the underlying cell biological features in the connectome. We apply NeuroSCAN to a complete published record of C. elegans brain neuropils and demonstrate how these integrated representations of neuronal relationships facilitate comparisons across connectomes, catalyzing new insights on the structure-function of the circuits and their changes during development. NeuroSCAN is designed for intuitive examination and comparisons across connectomes, enabling synthesis of knowledge from high-level abstractions of neuronal relationships derived from data science techniques to the detailed identification of the cell biological features underpinning these abstractions.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">arXiv</abbr></div> <div id="afrasiyabi_latent_2024" class="col-sm-8"> <div class="title">Latent Representation Learning for Multimodal Brain Activity Translation</div> <div class="author"> Arman Afrasiyabi, <em>Dhananjay Bhaskar</em>, Erica L. Busch, Laurent Caplette, Rahul Singh, Guillaume Lajoie, Nicholas B. Turk-Browne, and Smita Krishnaswamy</div> <div class="periodical"> Sep 2024 </div> <div class="periodical"> arXiv:2409.18462 [cs, q-bio] </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>Neuroscience employs diverse neuroimaging techniques, each offering distinct insights into brain activity, from electrophysiological recordings such as EEG, which have high temporal resolution, to hemodynamic modalities such as fMRI, which have increased spatial precision. However, integrating these heterogeneous data sources remains a challenge, which limits a comprehensive understanding of brain function. We present the Spatiotemporal Alignment of Multimodal Brain Activity (SAMBA) framework, which bridges the spatial and temporal resolution gaps across modalities by learning a unified latent space free of modality-specific biases. SAMBA introduces a novel attention-based wavelet decomposition for spectral filtering of electrophysiological recordings, graph attention networks to model functional connectivity between functional brain units, and recurrent layers to capture temporal autocorrelations in brain signal. We show that the training of SAMBA, aside from achieving translation, also learns a rich representation of brain information processing. We showcase this classify external stimuli driving brain activity from the representation learned in hidden layers of SAMBA, paving the way for broad downstream applications in neuroscience research and clinical contexts.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">arXiv</abbr></div> <div id="afrasiyabi_looking_2024" class="col-sm-8"> <div class="title">Looking through the mind’s eye via multimodal encoder-decoder networks</div> <div class="author"> Arman Afrasiyabi, Erica Busch, Rahul Singh, <em>Dhananjay Bhaskar</em>, Laurent Caplette, Nicholas Turk-Browne, and Smita Krishnaswamy</div> <div class="periodical"> Sep 2024 </div> <div class="periodical"> arXiv:2410.00047 [cs, eess, q-bio] </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>In this work, we explore the decoding of mental imagery from subjects using their fMRI measurements. In order to achieve this decoding, we first created a mapping between a subject’s fMRI signals elicited by the videos the subjects watched. This mapping associates the high dimensional fMRI activation states with visual imagery. Next, we prompted the subjects textually, primarily with emotion labels which had no direct reference to visual objects. Then to decode visual imagery that may have been in a person’s mind’s eye, we align a latent representation of these fMRI measurements with a corresponding video-fMRI based on textual labels given to the videos themselves. This alignment has the effect of overlapping the video fMRI embedding with the text-prompted fMRI embedding, thus allowing us to use our fMRI-to-video mapping to decode. Additionally, we enhance an existing fMRI dataset, initially consisting of data from five subjects, by including recordings from three more subjects gathered by our team. We demonstrate the efficacy of our model on this augmented dataset both in accurately creating a mapping, as well as in plausibly decoding mental imagery.</p> </div> </div> </div> </li> </ol> <h2 class="bibliography">2023</h2> <ol class="bibliography"><li> <div class="row"> <div class="col-sm-2 abbr"><abbr class="badge">arXiv</abbr></div> <div id="bhaskar_graph_2023" class="col-sm-8"> <div class="title">Graph topological property recovery with heat and wave dynamics-based features on graphs</div> <div class="author"> <em>Dhananjay Bhaskar</em>, Yanlei Zhang, Charles Xu, Xingzhi Sun, Oluwadamilola Fasina, Guy Wolf, Maximilian Nickel, Michael Perlmutter, and Smita Krishnaswamy</div> <div class="periodical"> Sep 2023 </div> <div class="periodical"> arXiv:2309.09924 [cs, eess, stat] </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> </div> <div class="abstract hidden"> <p>In this paper, we propose Graph Differential Equation Network (GDeNet), an approach that harnesses the expressive power of solutions to PDEs on a graph to obtain continuous node- and graph-level representations for various downstream tasks. We derive theoretical results connecting the dynamics of heat and wave equations to the spectral properties of the graph and to the behavior of continuous-time random walks on graphs. We demonstrate experimentally that these dynamics are able to capture salient aspects of graph geometry and topology by recovering generating parameters of random graphs, Ricci curvature, and persistent homology. Furthermore, we demonstrate the superior performance of GDeNet on real-world datasets including citation graphs, drug-like molecules, and proteins.</p> </div> </div> </div> </li></ol> </div> </article> </div> </div> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@4/imagesloaded.pkgd.min.js"></script> <script defer src="/assets/js/masonry.js" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.0.8/dist/medium-zoom.min.js" integrity="sha256-7PhEpEWEW0XXQ0k6kQrPKwuoIomz8R8IYyuU1Qew4P8=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?7b30caa5023af4af8408a472dc4e1ebb"></script> <script defer src="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.js"></script> <script src="/assets/js/no_defer.js?d633890033921b33e0ceb13d22340a9c"></script> <script defer src="/assets/js/common.js?acdb9690d7641b2f8d40529018c71a01"></script> <script defer src="/assets/js/copy_code.js?c9d9dd48933de3831b3ee5ec9c209cac" type="text/javascript"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script type="text/javascript">window.MathJax={tex:{tags:"ams"}};</script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.js"></script> <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script type="text/javascript">function progressBarSetup(){"max"in document.createElement("progress")?(initializeProgressElement(),$(document).on("scroll",function(){progressBar.attr({value:getCurrentScrollPosition()})}),$(window).on("resize",initializeProgressElement)):(resizeProgressBar(),$(document).on("scroll",resizeProgressBar),$(window).on("resize",resizeProgressBar))}function getCurrentScrollPosition(){return $(window).scrollTop()}function initializeProgressElement(){let e=$("#navbar").outerHeight(!0);$("body").css({"padding-top":e}),$("progress-container").css({"padding-top":e}),progressBar.css({top:e}),progressBar.attr({max:getDistanceToScroll(),value:getCurrentScrollPosition()})}function getDistanceToScroll(){return $(document).height()-$(window).height()}function resizeProgressBar(){progressBar.css({width:getWidthPercentage()+"%"})}function getWidthPercentage(){return getCurrentScrollPosition()/getDistanceToScroll()*100}const progressBar=$("#progress");window.onload=function(){setTimeout(progressBarSetup,50)};</script> </body> </html>